google/cloud/bigquery/opentelemetry_tracing.py

```diff
@@ -0,0 +1,115 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from contextlib import contextmanager
+from google.api_core.exceptions import GoogleAPICallError
+
+Logger = logging.getLogger(__name__)
+try:
+    from opentelemetry import trace
+    from opentelemetry.trace.status import Status
+    from opentelemetry.instrumentation.utils import http_status_to_canonical_code
+
+    HAS_OPENTELEMETRY = True
+
+except ImportError:
+    Logger.info(
+        "This service is instrumented using opentelemetry."
+        "Opentelemetry could not be imported, please"
+        "add opentelemetry-api and opentelemetry-instrumentation"
+        "packages, in order to get Big Query Tracing data"
+    )
+
+    HAS_OPENTELEMETRY = False
+
+_default_attributes = {
+    "db.system": "BigQuery",
+}
+
+
+@contextmanager
+def create_span(name, attributes=None, client=None, job_ref=None):
+    """Creates a ContextManager for a Span to be exported to the configured exporter. If no configuration
+        exists yields None.
+
+            Args:
+                name (str): Name that will be set for the span being created
+                attributes(Optional[dict]):
+                    Additional attributes that pertain to
+                    the specific API call (i.e. not a default attribute)
+                client (Optional[google.cloud.bigquery.client.Client]):
+                    Pass in a Client object to extract any attributes that may be
+                    relevant to it and add them to the created spans.
+                job_ref(Optional[google.cloud.bigquery.job._AsyncJob])
+                    Pass in a _AsyncJob object to extract any attributes that may be
+                    relevant to it and add them to the created spans.
+
+            Yields:
+                opentelemetry.trace.Span: Yields the newly created Span.
+
+            Raises:
+                google.api_core.exceptions.GoogleAPICallError:
+                    Raised if a span could not be yielded or issue with call to
+                    OpenTelemetry.
+            """
+    final_attributes = _get_final_span_attributes(attributes, client, job_ref)
+    if not HAS_OPENTELEMETRY:
+        yield None
+        return
+    tracer = trace.get_tracer(__name__)
+
+    # yield new span value
+    with tracer.start_as_current_span(name=name, attributes=final_attributes) as span:
+        try:
+            yield span
+            span.set_status(Status(http_status_to_canonical_code(200)))
+        except GoogleAPICallError as error:
+            if error.code is not None:
+                span.set_status(Status(http_status_to_canonical_code(error.code)))
+            raise
+
+
+def _get_final_span_attributes(attributes=None, client=None, job_ref=None):
+    if client:
+        client_attributes = _set_client_attributes(client)
+        _default_attributes.update(client_attributes)
+    elif job_ref:
```

Comment:
Would it be possible for someone to pass in both `client` and `job_ref`? (Should this just be `if` instead of `elif`?)

Reply:
Should not be possible since I instrumented this across two different modules. The only calls being made send a job_ref or client exclusively.

Reply:
Since the job classes do hold a client reference, I think it's reasonable to assume someone may think to set both. `create_span` could error in this case, but I think allowing it is preferable.

If both were passed in, I think preferring the `job_ref` properties is the right thing to do (that is, updating after the client attributes are set).

```suggestion
    if job_ref:
```